### ISSUE TITLE
FlashSelfAttention and SelfAttention in flash_attn.modules.mha give different results

### ISSUE THREAD

**Issue by senxiu-puleya** [Dao-AILab/flash-attention#1300]
Hello, I'm testing the attention functions before implementing them into my model, but I found that the modules `FlashSelfAttention` and `SelfAttention` in the demo give different results for the same input (and the result from `FlashSelfAttention` is wired). 
I have trace down to `flash_attn_cuda.fwd()` in `_flash_attn_forward` and everything seems to be simple and correct. Am I making a basic mistake, or is there something else at play?
Thanks very much !

python: 3.12.7
torch: 2.4.0
flash_attn: 2.6.3

ran in jupyter notebook

```python
import torch
from flash_attn.modules.mha import FlashSelfAttention, SelfAttention
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
q = torch.tensor([[[[0.1, 0.1, 0.1, 0.1]], [[0.1, 0.1, 0.1, 0.1]], [[0.3, 0.3, 0.3, 0.3]], [[0.1, 0.1, 0.1, 0.1]]]]).half()
k = torch.tensor([[[[0.1, 0.1, 0.1, 0.1]], [[0.1, 0.1, 0.1, 0.1]], [[0.3, 0.3, 0.3, 0.3]], [[0.1, 0.1, 0.1, 0.1]]]]).half()
v = torch.tensor([[[[0.1, 0.1, 0.1, 0.1]], [[0.1, 0.1, 0.1, 0.1]], [[0.3, 0.3, 0.3, 0.3]], [[0.1, 0.1, 0.1, 0.1]]]]).half()
qkv = torch.stack([q, k, v], dim=2).to(device)
q.shape, qkv.shape  # (B, S, H, D) & (B, S, 3, H, D)

# (torch.Size([1, 4, 1, 4]), torch.Size([1, 4, 3, 1, 4])) 
```

```python
fsa = FlashSelfAttention(causal=False, deterministic=True).to(device)
sa = SelfAttention(causal=False).to(device)
fsa.eval()
sa.eval()
fsa, sa

# (FlashSelfAttention(
#    (drop): Dropout(p=0.0, inplace=False)
#  ),
#  SelfAttention(
#    (drop): Dropout(p=0.0, inplace=False)
#  ))
```
```python
fsa(qkv)

# tensor([[[[ 0.1415,  0.1415,  0.1415,  0.1415]],
# 
#          [[-0.3311,  0.4624, -0.3311,  0.4624]],
# 
#          [[ 0.3000,  0.3000,  0.3000,  0.3000]],
# 
#          [[ 0.0989,  0.1006,  0.0989,  0.1006]]]], device='cuda:0',
#        dtype=torch.float16) 

```
```python
sa(qkv)

# tensor([[[[0.1515, 0.1515, 0.1515, 0.1515]],
# 
#          [[0.1515, 0.1515, 0.1515, 0.1515]],
# 
#          [[0.1547, 0.1547, 0.1547, 0.1547]],
# 
#          [[0.1515, 0.1515, 0.1515, 0.1515]]]], device='cuda:0',
#        dtype=torch.float16)
```

**Comment by tridao** [comment]
Looks like a bug as we recently changed the way we pad head dimensions that are not multiple of 8. Will investigate.

**Comment by tridao** [comment]
Did you install from the latest in `main` or from PyPI?

**Comment by senxiu-puleya** [comment]
> Did you install from the latest in `main` or from PyPI?
Hi, thanks for your reply. 
I'm using Windows and failed to build FlashAttention from PyPl/the main Git repository, so I'm using the pre-built wheel from https://github.com/bdashore3/flash-attention. I'll test it again once I've successfully installed it on Windows or have access to a Linux version.

**Comment by senxiu-puleya** [comment]
> Did you install from the latest in `main` or from PyPI?

Hi, I have successfully installed `flash_attn` through pip, and now there are only small differences in the output values between the `flash_attn` and normal attention calculations.

```python
# fsa
tensor([[[[0.1531, 0.1531, 0.1531, 0.1531]],

         [[0.1531, 0.1531, 0.1531, 0.1531]],

         [[0.1595, 0.1595, 0.1595, 0.1595]],

         [[0.1531, 0.1531, 0.1531, 0.1531]]]], device='cuda:0',
       dtype=torch.float16)

# sa
tensor([[[[0.1515, 0.1515, 0.1515, 0.1515]],

         [[0.1515, 0.1515, 0.1515, 0.1515]],

         [[0.1547, 0.1547, 0.1547, 0.1547]],

         [[0.1515, 0.1515, 0.1515, 0.1515]]]], device='cuda:0',
       dtype=torch.float16)
```
Are these differences within the acceptable error range? 
In any case, thank you for creating this wonderful library!

**Comment by tridao** [comment]
If I understand correctly the Windows wheels were built with `-DFLASHATTENTION_DISABLE_UNEVEN_K` (to reduce compile time):
https://github.com/bdashore3/flash-attention/blob/e8c161b31cd1c96901ecb29281281502c34089b9/setup.py#L288

This means it will only work with head dim = 32, 64, 96, 128, ..., 256. Your example has head dim = 4 so it won't work.
The code does support all head dim up to 256, one just needs to compile / use a wheel that does not disable "uneven K".
